From 5c377b241c01cf2f334b8e72b23435891d0a8574 Mon Sep 17 00:00:00 2001
From: Srikanth Uyyala <suyyala@codeaurora.org>
Date: Wed, 14 Mar 2018 10:42:10 +0530
Subject: [PATCH 053/100] msm: isp-legacy: Clear irq status if irq is set again
 with same value

Sometimes same irq is received even after clearing irq.
make sure irq status is rest by clearing the irq again.

Change-Id: Ib5355a4b9952fa7f02d1dea69b5fb767928e4b9f
Signed-off-by: Srikanth Uyyala <suyyala@codeaurora.org>
Signed-off-by: Lau <laststandrighthere@gmail.com>
Signed-off-by: Albert I <kras@raphielgang.org>
---
 .../platform/msm/camera_v2-legacy/isp/msm_isp47.c      | 10 ++++++++++
 1 file changed, 10 insertions(+)

diff --git a/drivers/media/platform/msm/camera_v2-legacy/isp/msm_isp47.c b/drivers/media/platform/msm/camera_v2-legacy/isp/msm_isp47.c
index ac655a6788ab..6671bdb294d1 100644
--- a/drivers/media/platform/msm/camera_v2-legacy/isp/msm_isp47.c
+++ b/drivers/media/platform/msm/camera_v2-legacy/isp/msm_isp47.c
@@ -555,6 +555,7 @@ void msm_vfe47_process_error_status(struct vfe_device *vfe_dev)
 void msm_vfe47_read_irq_status_and_clear(struct vfe_device *vfe_dev,
 	uint32_t *irq_status0, uint32_t *irq_status1)
 {
+	uint32_t count = 0;
 	*irq_status0 = msm_camera_io_r(vfe_dev->vfe_base + 0x6C);
 	*irq_status1 = msm_camera_io_r(vfe_dev->vfe_base + 0x70);
 	/* Mask off bits that are not enabled */
@@ -564,6 +565,15 @@ void msm_vfe47_read_irq_status_and_clear(struct vfe_device *vfe_dev,
 	*irq_status0 &= vfe_dev->irq0_mask;
 	*irq_status1 &= vfe_dev->irq1_mask;
 
+	/* check if status register is cleared if not clear again*/
+	while (*irq_status0 &&
+		(*irq_status0 & msm_camera_io_r(vfe_dev->vfe_base + 0x6C)) &&
+		(count < MAX_RECOVERY_THRESHOLD)) {
+		msm_camera_io_w(*irq_status0, vfe_dev->vfe_base + 0x64);
+		msm_camera_io_w_mb(1, vfe_dev->vfe_base + 0x58);
+		count++;
+	}
+
 	if (*irq_status1 & (1 << 0)) {
 		vfe_dev->error_info.camif_status =
 		msm_camera_io_r(vfe_dev->vfe_base + 0x4A4);
-- 
2.20.1

